### PR TITLE
Strict dimension checks for layer input tensors

### PR DESCRIPTION
**Pull request by timmoon10**
Currently, many of our layers require input tensors to have identical dimensions, e.g. binary math layers or some loss layers. This can get annoying since, for example, you can't directly compute the cross entropy between a 1000x1x1-tensor and a 1000-tensor. This PR relaxes the requirement to only require that input tensors have the same number of entries. I'm personally ambivalent about this change since you can now add a 2x3x4-tensor and a 6x2x2-tensor without any warning.

For reviewing, it will be helpful to know that the dimension checks for the categorical accuracy, cross entropy, mean absolute error, mean squared error, top-k categorical accuracy, binary math, and covariance layers are identical. The dimension checks for the Hadamard, sum, and weighted sum layers are also identical.

### CODE REVIEW

**Comment by timmoon10** [comment]
I'm wary of duplicating intricate dimension-checking over 10 layers (or more). Perhaps we should start thinking of proper tensor broadcasting semantics.

**Comment by ndryden** [comment]
I the mean time, could we just write a function (or two) that does the appropriate checks? We already have the check/error code duplicated.

**Comment by timmoon10** [comment]
Thinking more about it, do we even want this behavior? Following Numpy's broadcasting rules (https://docs.scipy.org/doc/numpy-1.15.0/user/basics.broadcasting.html), broadcasting a 1000x1x1-tensor with a 1000-tensor results in 1000x1x1000-tensors. We can add a special case for this particular situation, but what happens when we want to provide general tensor broadcasting?

**Comment by ndryden** [comment]
So, that's a good point. I fixed the issue I ran into via a reshape layer. PyTorch, for example, is actually even more strict than we are, in that it requires explicitly reshaping the output of a convolutional layer before passing it to a fully-connected layer. On the other hand, its loss functions automatically handle multi-dimensional inputs appropriately. Perhaps making it clear that everything is really tensors, and you have to handle that is worth-while. Certainly, it should be reasonable to practitioners (and if any practitioner _isn't_ aware that it's all a tensor, ...)

**Comment by timmoon10** [comment]
I've modified this PR to perform strict dimension checks. Merging it won't have any effect on behavior, but it does standardize the dimension checking code.
